client/src/indexer/mod.rs

```diff
@@ -1,57 +1,172 @@
-use std::{fmt::Debug, future::Future};
+use num_bigint::BigUint;
+use solana_sdk::signature::Keypair;
+use std::fmt::Debug;
 
-use light_concurrent_merkle_tree::light_hasher::Poseidon;
-use light_indexed_merkle_tree::{
-    array::{IndexedArray, IndexedElement},
-    reference::IndexedMerkleTree,
+use crate::rpc::RpcConnection;
+use account_compression::initialize_address_merkle_tree::{
+    Error as AccountCompressionError, Pubkey,
 };
+use light_compressed_token::TokenData;
+use light_hash_set::HashSetError;
+use light_hasher::Poseidon;
+use light_indexed_merkle_tree::array::{IndexedArray, IndexedElement};
+use light_indexed_merkle_tree::reference::IndexedMerkleTree;
 use light_merkle_tree_reference::MerkleTree;
-use light_sdk::{
-    compressed_account::CompressedAccountWithMerkleContext, event::PublicTransactionEvent,
-    proof::ProofRpcResult, token::TokenDataWithMerkleContext,
-};
-use num_bigint::BigUint;
-use solana_sdk::pubkey::Pubkey;
+use light_system_program::invoke::processor::CompressedProof;
+use light_system_program::sdk::compressed_account::CompressedAccountWithMerkleContext;
+use light_system_program::sdk::event::PublicTransactionEvent;
+use photon_api::apis::{default_api::GetCompressedAccountProofPostError, Error as PhotonApiError};
 use thiserror::Error;
 
-use crate::rpc::RpcConnection;
+#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
+pub struct StateMerkleTreeAccounts {
+    pub merkle_tree: Pubkey,
+    pub nullifier_queue: Pubkey,
+    pub cpi_context: Pubkey,
+}
 
-#[derive(Error, Debug)]
-pub enum IndexerError {
-    #[error("RPC Error: {0}")]
-    RpcError(#[from] solana_client::client_error::ClientError),
-    #[error("failed to deserialize account data")]
-    DeserializeError(#[from] solana_sdk::program_error::ProgramError),
-    #[error("failed to copy merkle tree")]
-    CopyMerkleTreeError(#[from] std::io::Error),
-    #[error("error: {0:?}")]
-    Custom(String),
-    #[error("unknown error")]
-    Unknown,
+#[derive(Debug, Clone, Copy)]
+pub struct AddressMerkleTreeAccounts {
+    pub merkle_tree: Pubkey,
+    pub queue: Pubkey,
+}
+
+#[derive(Debug, Clone)]
+pub struct StateMerkleTreeBundle {
+    pub rollover_fee: i64,
+    pub merkle_tree: Box<MerkleTree<Poseidon>>,
+    pub accounts: StateMerkleTreeAccounts,
+}
+
+#[derive(Debug, Clone)]
+pub struct AddressMerkleTreeBundle {
+    pub rollover_fee: i64,
+    pub merkle_tree: Box<IndexedMerkleTree<Poseidon, usize>>,
+    pub indexed_array: Box<IndexedArray<Poseidon, usize>>,
+    pub accounts: AddressMerkleTreeAccounts,
+}
+
+#[derive(Debug)]
+pub struct ProofRpcResult {
+    pub proof: CompressedProof,
+    pub root_indices: Vec<u16>,
+    pub address_root_indices: Vec<u16>,
+}
+
+#[derive(Debug, Clone)]
+pub struct TokenDataWithMerkleContext {
+    pub token_data: TokenData,
+    pub compressed_account: CompressedAccountWithMerkleContext,
 }
 
 pub trait Indexer<R: RpcConnection>: Sync + Send + Debug + 'static {
```

Comment:
do we even need the `Indexer` trait can't we just have a `photon-indexer` trait and a `test-indexer` trait?

Reply:
but indexers don't have to support methods like `get_address_merkle_trees_mut` these are `test-indexer` specific.
Essentially every method which is `unimplemented` by default can probably be moved to the test-indexer trait so that we get rid of all the `unimplemented` methods.
Do you think that would work?

Reply:
These are the methods I have in mind. Some already exist in the `MerkleTreeExt` trait already.
```
fn account_nullified(&mut self, _merkle_tree_pubkey: Pubkey, _account_hash: &str) {}

    fn address_tree_updated(
        &mut self,
        _merkle_tree_pubkey: Pubkey,
        _context: &NewAddressProofWithContext,
    ) {
    }

    fn get_state_merkle_tree_accounts(&self, _pubkeys: &[Pubkey]) -> Vec<StateMerkleTreeAccounts> {
        unimplemented!()
    }

    fn add_event_and_compressed_accounts(
        &mut self,
        _event: &PublicTransactionEvent,
    ) -> (
        Vec<CompressedAccountWithMerkleContext>,
        Vec<TokenDataWithMerkleContext>,
    ) {
        unimplemented!()
    }

    fn get_state_merkle_trees(&self) -> &Vec<StateMerkleTreeBundle> {
        unimplemented!()
    }

    fn get_state_merkle_trees_mut(&mut self) -> &mut Vec<StateMerkleTreeBundle> {
        unimplemented!()
    }

    fn get_address_merkle_trees(&self) -> &Vec<AddressMerkleTreeBundle> {
        unimplemented!()
    }

    fn get_address_merkle_trees_mut(&mut self) -> &mut Vec<AddressMerkleTreeBundle> {
        unimplemented!()
    }

    fn get_token_compressed_accounts(&self) -> &Vec<TokenDataWithMerkleContext> {
        unimplemented!()
    }

    fn get_payer(&self) -> &Keypair {
        unimplemented!()
    }

    fn get_group_pda(&self) -> &Pubkey {
        unimplemented!()
    }

    #[allow(async_fn_in_trait)]
    async fn create_proof_for_compressed_accounts(
        &mut self,
        _compressed_accounts: Option<&[[u8; 32]]>,
        _state_merkle_tree_pubkeys: Option<&[Pubkey]>,
        _new_addresses: Option<&[[u8; 32]]>,
        _address_merkle_tree_pubkeys: Option<Vec<Pubkey>>,
        _rpc: &mut R,
    ) -> ProofRpcResult {
        unimplemented!()
    }

    fn add_address_merkle_tree_accounts(
        &mut self,
        _merkle_tree_keypair: &Keypair,
        _queue_keypair: &Keypair,
        _owning_program_id: Option<Pubkey>,
    ) -> AddressMerkleTreeAccounts {
        unimplemented!()
    }

    fn get_compressed_accounts_by_owner(
        &self,
        _owner: &Pubkey,
    ) -> Vec<CompressedAccountWithMerkleContext> {
        unimplemented!()
    }

    fn get_compressed_token_accounts_by_owner(
        &self,
        _owner: &Pubkey,
    ) -> Vec<TokenDataWithMerkleContext> {
        unimplemented!()
    }

    fn add_state_bundle(&mut self, _state_bundle: StateMerkleTreeBundle) {
        unimplemented!()
    }
    ```